plugins/processors/topk/README.md

```diff
@@ -0,0 +1,97 @@
+# TopK Processor Plugin
+
+The TopK processor plugin is a filter designed to get the top series over a period of time. It can be tweaked to do its top k computation over a period of time, so spikes can be smoothed out.
+
+This plugin groups the metrics based on their name and tags, then generates aggregates values across each group base on fields selected by the user. It then sorts these groups based on these aggregations and returns any metric that belongs to a group in the top k (sorted by any of the aggregations). This means that when calculating the top k, more than k metrics may be returned.
+
+If only the very top k metrics are needed, regardless of grouping, the simple_topk setting will force each metric into its own individual group
+
+### Configuration:
+
+```toml
+[[processors.topk]]
+  ## How many seconds between aggregations
+  # period = 10
+
+  ## How many top metrics to return
+  # k = 10
+
+  ## Metrics are grouped based on their tags and name. The plugin aggregates
+  ## the selected fields of these groups of metrics and sorts the groups based
+  ## these aggregations
+
+  ## Over which tags should the aggregation be done. Globs can be specified, in
+  ## which case any tag matching the glob will aggregated over. If set to an
+  ## empty list is no aggregation over tags is done
+  # group_by = ['*']
+
+  ## Wheter or not to also group by metric name
+  # group_by_metric_name = true
+
+  ## The plugin can aggregate over several fields. If more than one field is
+  ## specified, an aggregation is calculated per group per field.
+
+  ## The plugin returns a metric if it's in a group in the top k groups,
+  ## ordered by any of the aggregations of the selected fields
+
+  ## This effectively means that more than K metrics may be returned. If you
+  ## need to return only the top k metrics regardless of grouping, use the simple_topk setting
```

Comment:
The processor goes through these steps:

1. Groups metrics in buckets using their tags & metrics name as key
2. Aggregates each of the selected fields for each bucket by some aggregation function (sum, mean, etc)
3. Orders the **buckets** by one of the generated aggregations, returns **all metrics in the top k** buckets, then reorders the buckets by the next of the generated aggregations, returns those metrics, etc, etc, etc, until it runs out of fields.
4. Finish the execution

So if you group by `PID` and aggregate over `cpu` and `memory` with a `K` of 10, you could get between 10 and 20 processes. This is also because PID's are unique, if I group by `datacenter` or `OS`, I could get **more** than 20 processes

This gives "complete" and "intuitive" results. What deficiencies do you see that make you look for an alternative?

Reply:
I think this is reasonable but there may be some confusion over why they are getting more than `k` results.

Reply:
Yeah, but the think about the problem is that unless you are doing a "simple" top k, you don't know before hand how many points of data you are gonna get.

Like if I made it so it only gives you the very top k of all the ingested metrics, some of the points could come in and out very rapidly, which would in a very unuseful graph. Maybe we just need make it clear in the documentation or change the name of the plugin.

Reply:
Okay, nevermind about this comment, I think how you have it is the best.